beam_jobs/base_model_validator.py

```diff
@@ -0,0 +1,239 @@
+        """
+        super(init)
+        self.regex_string = regex_string
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+          model: datastore_services.Model. Entity to validate.
+
+        Yields:
+          beam.pvalue.TaggedOutput. An element of the output PCollection for
+          the doFn which represents an error as a key value pair.
+        """
+        element = model.clone()
+        regex_string = self.regex_string
+
+        if not re.compile(regex_string).match(element.id):
+            yield beam.pvalue.TaggedOutput('error_category_id_check', (
+                'model %s' % ERROR_CATEGORY_ID_CHECK,
+                'Entity id %s: Entity id does not match regex pattern' % (
+                    element.id)
+            ))
+        yield element
+
+
+class ValidateDeleted(beam.DoFn):
+    """DoFn to check whether models marked for deletion are stale."""
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+          model: datastore_services.Model. Entity to validate.
+
+        Yields:
+          beam.pvalue.TaggedOutput. An element of the output PCollection for
+          the doFn which represents an error as a key value pair.
+        """
+        element = model.clone()
+        date_now = datetime.datetime.utcnow()
+
+        date_before_which_models_should_be_deleted = (
+            date_now -
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED
+        )
+
+        period_to_hard_delete_models_in_days = (
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED.days)
+
+        if element.last_updated < date_before_which_models_should_be_deleted:
+            yield beam.pvalue.TaggedOutput('error_category_stale_check', (
+                'entity %s' % ERROR_CATEGORY_STALE_CHECK,
+                'Entity id %s: model marked as deleted is older than %s weeks'
+                % (element.id, python_utils.divide(
+                    period_to_hard_delete_models_in_days, 7))
+            ))
+        yield element
+
+
+class ValidateModelTimeFields(beam.DoFn):
+    """DoFn to check whether created_on and last_updated timestamps are
+    valid."""
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+          model: datastore_services.Model. Entity to validate.
+
+        Yields:
+          beam.pvalue.TaggedOutput. An element of the output PCollection for
+          the doFn which represents an error as a key value pair.
+        """
+
+        element = model.clone()
+        if element.created_on > (element.last_updated
+                                 + datetime.timedelta(seconds=1)):
+            yield beam.pvalue.TaggedOutput(
+                'error_category_time_field_check', (
```

Comment:
Hmm, let's make and return a `ValidationError` class instead (strings feel too brittle and easy to get wrong): https://gist.github.com/brianrodri/ce2adb87e0a37bfe213a166658e8d6be